distributed/utils.py

```diff
@@ -1399,3 +1399,23 @@ def deserialize_for_cli(data):
         The de-serialized data
     """
     return json.loads(base64.urlsafe_b64decode(data.encode()).decode())
+
+
+class LRU(OrderedDict):
+    """ Limited size mapping, evicting the least recently looked-up key when full
+    """
+
+    def __init__(self, maxsize):
+        self.maxsize = maxsize
+        super().__init__()
+
+    def __getitem__(self, key):
+        value = super().__getitem__(key)
+        self.move_to_end(key)
+        return value
+
+    def __setitem__(self, key, value):
+        if len(self) >= self.maxsize:
+            oldest = next(iter(self))
+            del self[oldest]
```

Comment:
Can use `self.popitem(last=False)` here instead.

Reply:
Hmm I initially thought that would work, but I'm getting a `test_lru` failure locally (traceback below). It looks like `popitem` internally calls `__getitem__` which we're overloading here to include a call to `move_to_end`, which how some methods are implemented in `OrderedDict`.

I just pushed a commit which updates `LRU` to be a subclass of [`UserDict`](https://docs.python.org/3.8/library/collections.html#userdict-objects) and sets the underlying `.data` mapping to be an `OrderedDict`. This lets us still keep track or the order items are used, but avoids breaking other mapping methods like `LRU.popitem`

<details>
<summary>test_lru traceback:</summary>

```
============================================================ FAILURES ============================================================
____________________________________________________________ test_lru ____________________________________________________________

    def test_lru():

        l = LRU(maxsize=3)
        l["a"] = 1
        l["b"] = 2
        l["c"] = 3
        assert list(l.keys()) == ["a", "b", "c"]

        # Use "a" and ensure it becomes the last recently used item
        l["a"]
        assert list(l.keys()) == ["b", "c", "a"]

        # Ensure maxsize is respected
>       l["d"] = 4

distributed/tests/test_utils.py:617:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
distributed/utils.py:1427: in __setitem__
    self.popitem(last=False)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = LRU([('c', 3), ('a', 1)]), key = 'b'

    def __getitem__(self, key):
        value = super().__getitem__(key)
>       self.move_to_end(key)
E       KeyError: 'b'

distributed/utils.py:1414: KeyError
```

</details>